src/experiments/tcpsocket/api.js

```diff
@@ -0,0 +1,91 @@
+        return {
+            experiments: {
+                tcpsocket: {
+                    /** 
+                     * Send a DNS query stored in buf over a TCP socket to a 
+                     * nameserver addressed by addr
+                     */
+                    async sendDNSQuery(addr, buf) {
+                        let tcp_socket;
+
+                        try {
+                            // Wait until the socket is open before sending data.
+                            // If we get an 'error' event before an 'open' event, 
+                            // throw an ExtensionError.
+                            tcp_socket = new TCPSocket(addr, 53, { binaryType: "arraybuffer" });
+                            let eventType = await new Promise((resolve, reject) => {
+                                tcp_socket.onopen = ((event) => {
+                                    resolve(event.type);
+                                });
+
+                                tcp_socket.onerror = ((event) => {
+                                    throw new ExtensionError(event.name);
+                                });
+                            });
+                            if (eventType != "open" || tcp_socket.readyState != "open") {
+                                throw new ExtensionError("TCP socket didn't properly open");
+                            }
+
+                            // After we know that the socket is open, send the bytes
+                            tcp_socket.send(buf.buffer, buf.byteOffset, buf.byteLength);
+                            let responseBytes = await new Promise((resolve, reject) => {
+                                let data = new Uint8Array();
+                                let expectedLength;
+
+                                tcp_socket.ondata = ((event) => {
+                                    if (data.length == 0) {
+                                        expectedLength = new DataView(event.data).getUint16(0);
+                                        let receivedData = new Uint8Array(event.data);
+                                        data = receivedData;
+                                    } else {
+                                        data = concatUint8Arrays(data, receivedData);
+                                    }
+
+                                    // Check if we have got all the expected data, or if we've got too much data
+                                    if (data.length == expectedLength + 2) {
+                                        resolve(data);
+                                    } else if (data.length > expectedLength + 2) {
+                                        throw new ExtensionError("Got too many bytes from TCP");
+                                    }
+                                });
+
+                                tcp_socket.onopen = ((event) => {
+                                    throw new ExtensionError("Got an additional open event");
```

Comment:
None of these `throw new ExtensionError` will propagate to the caller of `sendDNSQuery`.

If you want to know the linear execution flow, this is what happens:

1. `sendDNSQuery` is called.
2. ...
3. Create `tcp_socket` + `tcp_socket.onopen` listener is registered.
4. `Promise` is created.
5. Yielding to event loop (due to `await`).
6. JavaScript execution continues (up the call stack).
7. At some point, the `open` event is dispatched.
8. If your handler throws an error, then the (native) caller that dispatched the "open" event may or may not have a chance of handling the error (depending on the implementation). In any case, your `sendDNSQuery` function will NOT receive this error.
9. If your handler calls `resolve()` or `reject()`, then eventually execution will resume from where we left of at step 5.

I think that you expected step 8 to pass the `ExtensionError` to the `sendDNSQuery` caller, but that is not the case. If you really want to pass on an error, use `reject()` instead of `throw`.

Reply:
I think I addressed this in aa004e3